streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -213,12 +213,15 @@ public boolean isValidTransition(final State newState) {
     private final Object stateLock = new Object();
     protected volatile State state = State.CREATED;
 
-    private boolean waitOnState(final State targetState, final long waitMs) {
+    private boolean waitOnState(final State targetState, final long waitMs, final boolean newSemantics) {
         final long begin = time.milliseconds();
         synchronized (stateLock) {
             long elapsedMs = 0L;
             while (state != targetState) {
                 if (waitMs == 0) {
+                    if (newSemantics)
+                        return false;
```

Comment:
Nit: every conditional needs to have braces (per the code style)

not nit: I find this logic a little difficult to follow. Contrary to what @mjsax suggested, wouldn't it be pretty straightforward to map the old semantics on to the new ones like this:
* negative numbers => 0
* 0 => Long.MAX_VALUE
* all other arguments stay the same
?

Then, the old close method could just transform its arguments and call the new method, with no need to have this "new semantics" flag and an early return in the middle of the loop.

Reply:
@vvcephei I like your proposal. Thanks!
Changed PR according to it.